### ISSUE TITLE
(custom resources): Can not get public key for a KMS key

### ISSUE THREAD

**Issue by sashee** [aws/aws-cdk#19065]
### What is the problem?

I tried to extract the public key for an asymmetric KMS key but I get a ```Response is not valid JSON``` error.

### Reproduction Steps

```ts
const key = new aws_kms.Key(
  this,
  "key",
  {
    keySpec: aws_kms.KeySpec.ECC_NIST_P384,
    keyUsage: aws_kms.KeyUsage.SIGN_VERIFY,
  }
);
const publicKey = new custom_resources.AwsCustomResource(
  this,
  "publicKey",
  {
    onCreate: {
      service: "KMS",
      action: "getPublicKey",
      parameters: {
        KeyId: key.keyArn,
      },
      physicalResourceId: custom_resources.PhysicalResourceId.of(key.keyArn),
      outputPaths: ['PublicKey'],
    },
    policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
      resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE
    }),
  }
);
```

### What did you expect to happen?

I expected that the stack deploys and I can extract the public key.

### What actually happened?

```Response is not valid JSON```

In the CloudWatch Logs, I see that the PublicKey is retrieved, but it seems in a unusable format:

![image](https://user-images.githubusercontent.com/82075/154947182-c371cf8a-e1cd-4369-b01c-da4ec20f1d34.png)

### CDK CLI Version

2.12.0

### Node.js Version

v16.14.0

### OS

Ubuntu

### Language

Typescript

**Comment by lacteolus** [comment]
The same issue with CDK 2.20.0.
Tried to create our own custom resource backed by custom lambda to get public key using boto3 but faced the problem described in https://github.com/aws-samples/aws-cdk-examples/discussions/641

**Comment by peterwoodworth** [comment]
I wonder if this is because the response is a [`blob`](https://docs.aws.amazon.com/kms/latest/APIReference/API_GetPublicKey.html#KMS-GetPublicKey-response-PublicKey)?

**Comment by skinny85** [comment]
Unfortunately, I'm not sure CDK can do much here - we don't really control wha the response from the services is.

**Comment by stefanopallicca-imagicle** [comment]
> Unfortunately, I'm not sure CDK can do much here - we don't really control wha the response from the services is.

Maybe it's a matter of _reformatting_ (I'm sorry if the word is not the proper one) the command output to a sort of human readable?

After all, the `aws kms get-public-key` cli command output has the expected format:

![image](https://user-images.githubusercontent.com/81425747/167367646-54af839f-8233-435f-872c-c30773cca4ea.png)

**Comment by skinny85** [comment]
The fact that you get that error message might also suggest that the response was an "Access Denied", which is not a JSON objcect.

Can you try temporarly giving the Custom Resource admin permissions, and see if that changes anything? (I guess also allow all principals from the account `"kms:*"` on the Key, just in case)

**Comment by stefanopallicca-imagicle** [comment]
@skinny85 thanks for the answer. I've tried what you asked for, but still receiving `Response is not valid JSON`

**Comment by johannes-sscrc** [comment]
The SDK returns [this object](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-kms/interfaces/getpublickeycommandoutput.html). Of particular interest is PublicKey, which is a Uint8Array (i.e. it's not a base64-encoded string).

**Comment by skinny85** [comment]
Interesting! That's probably the source of the `Response is not valid JSON` error.

**Comment by Jacco** [comment]
Think it is the response function in the lambda that implements the Custom Resource

```typescript
  function respond(responseStatus: string, reason: string, physicalResourceId: string, data: any) {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason: reason,
      PhysicalResourceId: physicalResourceId,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: false,
      Data: data,
    });
```

there is some code trying to deal with buffers:

```python
          const childKey = Buffer.isBuffer(child[key]) ? child[key].toString('utf8') : child[key];
```

but my guess is that this fails.

**Comment by johannes-sscrc** [comment]
I don't think this is what's causing the "not valid JSON" error, but I'd like to point out that decoding an arbitrary buffer with `.toString('utf8')` is potentially a lossy conversion. The code will try to decode the buffer as UTF-8, and any part of it that is not valid UTF-8 will be replaced by the Unicode replacement character � (U+FFFD), making it impossible to get the original buffer.

**Comment by dmitridr** [comment]
Had the same issue as everyone here.

The line @johannes-sscrc linked does seem to be related to the error. Indeed, Publickey seems to be in a binary format: DER, and the proper way to encode it appears to be base64 and not utf8. 

Thus, a quick fix for me was to replace the line you gave with:

```node
const childKey = Buffer.isBuffer(child[key]) ? child[key].toString('base64') : child[key];
```

Then the PublicKey is returned in base64. Generally, it would be good to have a hook/param here to allow the caller to specify how to encode the buffer for specific keys, in order to unblock such issues in the general case.

**Comment by synthetic-luis** [comment]
@sashee did you end up finding a work-around for getting the public key of a KMS key using a custom resource? I'm getting the same invalid JSON error.

**Comment by sashee** [comment]
@synthetic-luis , yeah, a custom resource can fetch the public key and output it for other resources. Unfortunately, I can't provide code example as it was done for a client, but the implementation was straightforward.

**Comment by synthetic-luis** [comment]
@sashee thanks for the response. Can you please elaborate on how you then overcame the invalid JSON issue (in general terms) as it relates to the code you posted above? For example, did you have to change your code above (e.g. add a new parameter to the `getPublicKey` call?)

**Comment by sashee** [comment]
@synthetic-luis , I just checked the code and I remember wrongly. So I needed a keypair for IVS playback key and I thought to use KMS for that. It did not work (partly because of the issue here) so I implemented a lambda function that generates the key. Not the ideal solution, but it works reliably so far.

**Comment by synthetic-luis** [comment]
@sashee this helps a lot. Thank you!

**Comment by DanielLaberge** [comment]
Would love to have a fix or workaround for this.

**Comment by synthetic-luis** [comment]
@DanielLaberge if it helps, what worked for me was creating a post deploy script that uses the KMS SDK to do what I needed to do

**Comment by aemada-aws** [comment]
Looks like a duplicate of https://github.com/aws/aws-cdk/issues/26730, please correct me if this is not fixed yet.
